dali/python/nvidia/dali/_debug_mode.py

```diff
@@ -0,0 +1,332 @@
+
+    def __ror__(self, other):
+        return _PipelineDebug.current()._wrap_op_call(_arithm_op, "bitor", other, self)
+
+    def __xor__(self, other):
+        return _PipelineDebug.current()._wrap_op_call(_arithm_op, "bitxor", self, other)
+
+    def __rxor__(self, other):
+        return _PipelineDebug.current()._wrap_op_call(_arithm_op, "bitxor", other, self)
+
+
+class _PipelineDebugRunner():
+    def __init__(self, pipeline):
+        self.pipeline = pipeline
+
+    def __enter__(self):
+        self.pipeline._debug_on = True
+        self.pipeline._external_source_debug = True
+        self.pipeline._cur_subpipeline_id = -1
+        return super(type(self.pipeline), self.pipeline).__enter__()
+
+    def __exit__(self, *exc):
+        self.pipeline._debug_on = False
+        self.pipeline._external_source_debug = False
+        if not self.pipeline._subpipelines_built:
+            self.pipeline._subpipelines_built = True
+        return super(type(self.pipeline), self.pipeline).__exit__(*exc)
+
+
+class _PipelineDebug(_pipeline.Pipeline):
+    """Debug mode for pipeline. Allows access to data inside the pipeline execution by wrapping all
+     operators inside their pipelines"""
+
+    def __init__(self, exec_func, **kwargs):
+        kwargs['debug'] = False
+        super().__init__(**kwargs)
+        kwargs['exec_pipelined'] = False
+        kwargs['exec_async'] = False
+        self._debug_on = False
+        self._external_source_debug = False
+        self._es_input_name = 'input_'
+        self._es_kwarg_name = 'kwarg_'
+        self._subpipeline_kwargs = kwargs
+        self._subpipelines = {}
+        self._external_source_pipelines = {}
+        self._feed_input_data = {}
+        self._subpipelines_built = False
+        self._cur_subpipeline_id = -1
+        self._exec_func = exec_func
+        self._runner = _PipelineDebugRunner(self)
+
+    def __enter__(self):
+        raise RuntimeError("Currently pipeline in debug mode works only with `pipeline_def` decorator."
+                           "Using `with` statement is not supported.")
+
+    def build(self):
+        """Build the pipeline.
+        
+        Refer to :meth:`Pipeline.build() <nvidia.dali.Pipeline.build>` for details."""
+        self._built = True
+
+    def run(self):
+        """Run the pipeline and return the result."""
+        import numpy as np
+        if not self._built:
+            raise RuntimeError('Pipeline must be built first.')
+        with self._runner:
+            res = self._exec_func()
+            if res is None:
+                res = ()
+            elif not isinstance(res, tuple):
+                res = (res,)
+
+            # Transforming all variables to TensorLists.
+            return tuple([val.get() if isinstance(val, DataNodeDebug) else _Tensors.TensorListCPU(
+                np.tile(val, (self._max_batch_size, *[1]*np.array(val).ndim))) for val in res])
```

Comment:
I assume this is to return the constants that were present in the original `return` statement, and would be captured by dali.fn.constant?

Reply:
That's right